manpage/mcxcontest.1
.\" Manpage for mcxcontest.
.\" Contact fangqq@gmail.com to correct errors or typos.
.TH man 7 "26 January 2025" "v2025.10" "mcxcontest man page"
.SH NAME
mcxcontest \- benchmark your NVIDIA GPU hardware using MCX built-in tests
.SH SYNOPSIS
mcxcontest <option1> <option2> ...
.SH DESCRIPTION
Monte Carlo eXtreme (mcx) is a CUDA-based GPU-accelerated 3D photon transportsimulator developed by Dr. Qianqian Fang at the Computational Optics andTranslational Imaging (COTI) Lab, at Northeastern University, Boston,
USA. More details regarding mcx can be found at http://mcx.space/

The `mcxcontest` script benchmarks your NVIDIA GPU using 3 built-in tests
in `mcx`, namely `cube60`, `cube60b` and `cube60planar`. By default, a totalof 10^8 photons will be simulated and the running speed, in photon/ms, foreach benchmark is reported. The final score of the GPU is the summation of
the 3 speed readings. The higher the score, the faster the GPU hardware.

This script is written in Perl. If utility `curl` is installed, it uses
`curl` to submit the hardware score to our database, should a user choose to
do so. If `curl` is not installed, mcxcontest requires `LWP::UserAgent` and`LWP::Simple` Perl modules to submit the user data.


.SH OPTIONS
The `mcxcontest` script can be executed without any parameter. This runs the
standard benchmark (10^8 photons) using the first GPU device (listd by `mcx -L`).

In addition, the following special parameters are supported:

.TP
\fB-l\fR
Use `-l` to download benchmark database from http://mcx.space/gpubench/ and compare
with your GPU results.

.TP
\fB-s\fR
Use `-s` to submit the benchmarks to our database at http://mcx.space/gpubench/

.TP
\fB-c\fR
Use `-c` to generate JSON formatted benchmark reports in the compact format.

.TP
\fB-o 'additional mcx parameters'\fR
Using `-o`, one can attach additional command line parameters to overwrite thedefault MCX benchmark parameters. Therefore, any standard parameters permitted byMCX is allowed to be used with this command.

.TP
\fB-L\fR
Use `-L` to list supported GPUs.

.TP
\fB-n <num>/-G <01>/-W <w1,w2,...>\fR
Use these parameters to adjust photon number (-n), which GPU to use (-G) and,
if multiple GPUs are used, ratios to split the workload (-W).

.TP
\fB--bin /path/to/mcx/binary\fR
Use --bin to manually specify the location of the mcx executable.

.TP
\fB--get URL_to_browse_database\fR
Use --get to set the database browse URL to a different location; the default MCX
GPU database URL is http://mcx.space/gpubench/

.TP
\fB--post URL_to_post_results_to_database\fR
Use `--post` to set the database submission URL to a different location; thedefault MCX GPU database URL is http://mcx.space/gpubench/gpucontest.cgi

.TP
\fB--help\fR
Print help information.


.SH SEE ALSO
mcx(7), mmc(7), mcxcl(7), mcxstudio(7), mcxshow(7), mcxviewer(7)
.SH AUTHOR
Qianqian Fang (q.fang@neu.edu)


utils/NEWS
= Change Log =

Major updates are marked with a "*"

== MCX v2025.6.pre (Kilo-Kelvin - 2.8), FangQ <q.fang (a) neu.edu> ==


== MCX v2025 (Jumbo Jolt - 2.6), FangQ <q.fang (a) neu.edu> ==

 2025-01-22 [24b445e] [bug] fix incorrect per-voxel pathlength when mua->0, #164
 2025-01-21 [1f536ba] [bug] fix windows -N error
 2025-01-21 [aeca212] [cmake] update cmake to add -N support
 2025-01-21 [134ab82] [ci] fix windows msvc compilation error
 2025-01-21 [12071fd] [feat] add -N/--net to download simulations from NeuroJSON.io, add -Q
 2025-01-10 [cf10d5f] [pmcx] Parse issavedet field of the cfg dict as an int instead of a bool.
 2025-01-03 [8f433e9] [feat] allow plotting logical arrays in mcxplotvol
 2024-12-15 [eafea84] [feat] polish and miss_hit format of demo_mcxlab_replay_traj.m
 2024-12-15 [82e100b] [feat] add missing demo_mcxlab_replay_traj.m script
 2024-12-11 [22bf12a] [bug] fix mcxlab castlist
 2024-12-04 [86dcba9] [bug] force most scalar inputs to be double to avoid incorrect typecasting
 2024-11-25 [4b7020d] [ci] upgrade pmcx macos build to macos-13
 2024-11-25 [51e9970] [bug] fix mcx crash when replay is requested without seed file
 2024-11-22 [e8907db] [ci] include vcomp140.dll in windows binary
 2024-11-22 [2902119] [ci] further test vcomp140
 2024-11-22 [c3d413d] [ci] find vcomp140.dll
 2024-11-21 [4f1335c] [ci] print vcomp1xx.dll path
 2024-11-21 [d033755] [bug] apply patch in #235 to pmcx, bump pmcx to v0.3.6
 2024-11-21 [744c088] [bug] correctly return integer based w0 and detid in pattern3d
 2024-11-21 [dcf11c4] [bug] enable photon sharing for pattern3d source
 2024-11-18 [bd62362] [bug] fix mcxlab crash with >2^31-1 voxels, fix #235, revert #164
 2024-11-13 [188338b] [feat] negative detector radius captures but does not save photons
 2024-11-12 [dfc704f] [ci] reduce macos binary size using strip
 2024-11-12 [7d33c14] [ci] disable octave download on macos
 2024-11-12 [43e7416] [ci] download octave app with -L
 2024-11-12 [c978a3e] [ci] use octave app on macos
 2024-11-12 [30cc3cb] [ci] upload macos mcx package
 2024-11-12 [24d1257] [ci] downgrade download-artifact as it does not support multiple jobs
 2024-11-10 [f17dcc6] [bug] fix mcxlab crash when using issave2pt=0 with photonsharing
 2024-11-09 [8c69db3] [ci] update setup.py for pmcx
 2024-11-09 [fd83829] [ci] disable CMAKE_RANLIB
 2024-11-09 [dbada19] [ci] debug ranlib
 2024-11-09 [208c31d] [ci] use verbose option to debug macos build flags
 2024-11-09 [4e0c672] [bug] fix maskdet 1cube test
 2024-11-08 [ee5be15] [feat] make onecube benchmark maskdet work, fix --dumpmask
 2024-11-08 [f2d3bc4] [feat] support 1x1x1 volume, add onecube/twocube benchmarks, det not working
 2024-11-06 [eaafb0b] [feat] disabling issaveref if issave2pt is false
 2024-10-14 [6dbcc0f] Merge pull request #234 from epini/epini-patch-1
 2024-10-09 [0411ec6] Fix data ordering in traj.iquv
 2024-09-25 [8c0cfec] [pmcx] bump up pmcx version to 0.3.5 after fixing #233
 2024-09-25 [58dec12] [bug] angleinvcdf and invcdf are not read in full in pmcx, fix #233
 2024-09-14 [7b8ecb6] [ci] restore macos-12
 2024-09-13 [6fbf3db] [bug] fix the potential typo in Custo et al for CSF mua, fix #232
 2024-09-03 [c9456a7] [ci] fix action alert related to download-artifact
 2024-08-22 [70f95ba] [pmcx] fix CI error for pmcx
 2024-08-22 [82fb8f1] [feat] allow mcxlab and pmcx to use cfg.flog to control log printing
 2024-08-18 [17e347c] [bug] fix lzma memory leakage, NeuroJSON/zmat#11, lloyd/easylzma#4
 2024-08-06 [5cc92ab] [bug] avoid using the same RNG seed when -E -1 on multiple GPUs
 2024-08-06 [3c480b6] [doc] clarify the default RNG seed (1648335518) in html doc
 2024-07-22 [d8959eb] [ci] update macos-11 to 12 as 11 no longer works
 2024-07-14 [b17cb1a] [bug] mcxplotshapes patch by ChenJY-L to plot Box, close #227
 2024-07-14 [78716e4] [bug] avoid overwriting mua/mus when one is nan, fix #224, fix #225
 2024-07-04 [80b5794] [bug] read g and n when mua or mus=nan in asgn_float, close #225, close #224
 2024-06-22 [f959c71] [bug] store reflection position in trajectory
 2024-06-19 [4ff5b60] [bug] print large photon numbers without overflow
 2024-06-11 [d66a0a3] [test] sync test script between mcx and mcxcl after fangq/mcxcl#57
 2024-06-11 [1e7d0f1] [bug] reset replay.tof when tof exceeds tend, like fangq/mcxcl#57
 2024-06-07 [026eebf] [bug] ensure to clear shared mem buffer regardless of weight, #222
 2024-06-06 [b41c915] [bug] fix skewed nscat distribution, fix #222
 2024-06-05 [654dff1] [bug] fix outputtype error in json2mcx, reformat with miss_hit
 2024-06-05 [25b0268] [feat] export iquv in trajectory data when cfg.istrajstokes=1
 2024-05-18 [0a76d17] [bug] fix mcxsvmc flipped normal direction in octave, fix #221
 2024-05-18 [2f42524] [ci] downgrade matlab from v2024a to v2022a
 2024-05-14 [56aa355] [doc] update jsonlan and neurojson toolbox download links
 2024-04-25 [ea67ea9] [bug] avoid double-base64-encoding when -Z 2 is used, fix #219
 2024-03-28 [3b7e11c] [doc] fix incorrect default value for gscatter
 2024-03-25 [b4706ae] [release] update winget mcxstudio package
 2024-03-18 [2953735] [doc] add documentation on the srcid output in detp and traj
 2024-03-18 [94961f3] [bug] return source ID in multi-source simulation, fix #217
 2024-03-17 [3c3d755] [release] post v2024.2 release action, close #216
 2024-03-17 [7902a4e] [mcxcloud] update docker image to v2024.2
 2024-03-15 [08bfe11] [feat] support _ArrayData_ in Shapes volume input

== MCX v2024.2 (Interstellar Ion - 2.2), FangQ <q.fang (a) neu.edu> ==

 2024-03-13 [abdee14] [bug] fix multi-source replay bug, close #215
 2024-03-11 [9250a0d] [doc] make final doc update, bump pmcx to v0.3.2
 2024-03-10 [4e7f404] [ci] revert to windows-2019, add help info for #214, add note on nvidia-uvm
 2024-03-10 [2750d70] [ci] choco install is failing on Windows, see actions/runner-images#9477
 2024-03-10 [7b6e0e0] [optimize] cut hyperboloid gaussian register use from 15 to 3, #127,#214

speedcontest/mcxcontest.pl
#!/usr/bin/perl
###############################################################################
#
# MCX Benchmark and Speed Contest
#
# Author:  Qianqian Fang <q.fang at neu.edu>
# License: GPLv3
# Version: 0.5
# URL:     http://mcx.space/gpubench
# Github:  https://github.com/fangq/mcx/
#
###############################################################################
#
# Dependencies:
#
# For Linux and Mac OS: no additional package is needed, Perl is supported by default
#
# For Windows: please install cygwin64 (http://cygwin.com/) or MSYS2 (http://msys2.org/)
#
# Specifically, this script requires Perl 5.x and curl (if curl is not install 
# in the path, mcxcontest.pl looks for the LWP::UserAgent and LWP::Simple perl modules.
#
# How to use:
#    type "mcxcontest.pl --help" to see the format
#
###############################################################################

use strict;
use warnings;
use Term::ANSIColor qw(:constants);

my $MCX="../bin/mcx";
my $URL="http://mcx.space/gpubench/";
my $POSTURL="http://mcx.space/gpubench/gpucontest.cgi";

my %report=();
my $prettyjson=1;

# parse commandline options
my $options='';
my $mcxopt='-n 1e8';
while(my $opt = $ARGV[0]) {
    if($opt =~ /^-[coslnGWL-]/){
        if($opt eq '-s'){
		$options.='s';
	}elsif($opt eq '-c'){
		$prettyjson=0;
	}elsif($opt eq '-l'){
		$options.='l';
	}elsif($opt eq '-L'){
		system("$MCX -L");
		exit 0;
	}elsif($opt eq '-o'){
		shift;
		$mcxopt=shift;
		next;
        }elsif($opt =~ /^-[nGW]$/){
                shift;
                $mcxopt.=" $opt ".shift;
                next;
	}elsif($opt eq '--bin'){
		shift;
		$MCX=shift;
		next;
	}elsif($opt eq '--get'){
		shift;
		$URL=shift;
		next;
	}elsif($opt eq '--post'){
		shift;
		$POSTURL=shift;
		next;
	}elsif($opt eq '--help'){
		printf("mcxcontest - running MCX benchmarks and submit to MCX Speed Contest
	Format: %s <option1> <option2> ...\n
The supported options include (multiple parameters can be used, separated by spaces)
	-s      submit result to MCX Speed Contest by default (will ask if not used)
        -c      print JSON in compact form, otherwise, print in the indented form
	-l      compare your GPU with other GPUs in the database (will ask if not used)
	-L      list all GPUs supported on your system
	-o 'mcx options'   supply additional mcx command line options, such as '-n 1e6'
	-n <num> / -G <01> / -W <w1,w2,..> additional mcx command options add after those in -o
	--bin /path/to/mcx  manually specify mcx binary location (default: ../bin/mcx)
	--get url  specify mcx benchmark web link (default: http://mcx.space/gpubench)
	--post url specify submission link (default: http://mcx.space/gpubench/gpucontest.cgi)\n",
		$0);
		exit 0;
	}else{
		die('invalid command line option');
	}
	shift;
    }else{
        die('invalid command line option');
    }
}

my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime(time);
$report{'date'} = sprintf("%.4d-%.2d-%.2d %.2d:%.2d:%.2d", $year+1900, $mon+1, $mday, $hour, $min, $sec);

## perform benchmark

&listgpu(\%report,'gpu',$mcxopt);
&runbench('cube60',\%report,'benchmark1',17.69,$mcxopt);
&runbench('cube60b',\%report,'benchmark2',27.24,$mcxopt);
&runbench('cube60planar',\%report,'benchmark3',25.51,$mcxopt);

## assemble report data

$report{'speedsum'}=$report{'benchmark1'}{'speed'}+$report{'benchmark2'}{'speed'}
    +$report{'benchmark3'}{'speed'};
$report{'version'}=1;
$report{'mcxversion'}=$report{'benchmark1'}{'mcxversion'};

print("Speed test completed\nYour device(s) score is ");
print(RED);
printf("%.0f\n",$report{'speedsum'});
print(RESET);

## download report and compare

my $ans='';

if(! ($options=~/l/)){
	print "Do you want to compare your score with other submitted results ([yes]/no)?\n";
	$ans=<STDIN>;
	chomp $ans; 
}
if($options=~/l/ || $ans =~/^yes/i || $ans eq ''){
	&comparegpu($URL);
}
$ans='';

## submit report
if(! ($options=~/s/)){
	print "\nDo you wish to submit your GPU benchmark data to our database (yes/[no])?\n";
	$ans=<STDIN>;
	chomp $ans; 
}
if($options=~/s/ || $ans =~/^yes/i){
	&submitresult(\%report,$POSTURL,$prettyjson);
}
$ans='';

#==============================================================================
sub listgpu(){
	my ($report, $keyname, $mcxopt)=@_;
	my $mcxoutput=`$MCX -L`;

	$mcxoutput=~s/\e\[\d+(?>(;\d+)*)m//g;

	printf("Running '$MCX -L' to inquire GPUs ...\n");

	if($mcxoutput =~ /Global Memory/i){
		my @gpustr=split(/=+\s+[CG]PU .*\s+=+/,$mcxoutput);
		foreach my $gpurec (@gpustr){
			my @gpudata=split(/\n/,$gpurec);
			my %gpuinfo=();
			foreach my $info (@gpudata){
				if($info =~ /Device (\d+) of (\d+)\s*:\s*(.*)$/){
					$gpuinfo{'name'}=$3;
					$gpuinfo{'id'}=$1+0;
					$gpuinfo{'devcount'}=$2+0;
				}elsif($info =~ /Compute Capability\s*:\s*(\d+)\.(\d+)/i){
					$gpuinfo{'major'}=$1+0;
					$gpuinfo{'minor'}=$2+0;
				}elsif($info =~ /Global Memory\s*:\s*(\d+)\s*B/i){
					$gpuinfo{'globalmem'}=$1+0;
				}elsif($info =~ /Constant Memory\s*:\s*(\d+)\s*B/i){
					$gpuinfo{'constmem'}=$1+0;
				}elsif($info =~ /Shared Memory\s*:\s*(\d+)\s*B/i){
					$gpuinfo{'sharedmem'}=$1+0;
				}elsif($info =~ /Registers\s*:\s*(\d+)$/){
					$gpuinfo{'regcount'}=$1+0;
				}elsif($info =~ /Clock Speed\s*:\s*([0-9.]+)\s+GHz/i){
					$gpuinfo{'clock'}=$1*1e6;
				}elsif($info =~ /Number of Cores\s*:\s*(\d+)$/){
					$gpuinfo{'core'}=$1+0;
				}elsif($info =~ /Number of SM.*\s*:\s*(\d+)$/){
					$gpuinfo{'sm'}=$1+0;
				}elsif($info =~ /Auto-thread\s*:\s*(\d+)$/){
					$gpuinfo{'autothread'}=$1+0;
				}elsif($info =~ /Auto-block\s*:\s*(\d+)$/){
					$gpuinfo{'autoblock'}=$1+0;
				}
			}
			if(%gpuinfo){
				push(@{$report{$keyname}}, \%gpuinfo);
				if(!($mcxopt=~/-G\s+["']*[0-9]+['"]*/ || $mcxopt=~/--gpu\s+['"]*[0-9]+['"]*/)){
					last;
				}
			}
		}
		if(length($report{$keyname})==0){
			die('no GPU found on your system, can not continue');
		}
	}else{
                die('GPU listing failed, output incomplete');
        }
	if($mcxopt=~/-G\s+['"]*([0-9]+)['"]*/ || $mcxopt=~/--gpu\s+['"]*([0-9]+)['"]*/){
		if($1 ne '' && $1>0){
			my @newgpu=();
			my $mask=$1;
			my $count=0;
			if($mask =~/^[01]+$/){
				foreach my $i (split(//,$mask)){
					if($count>=@{$report{$keyname}}){
						last;
					}
					push(@newgpu,$report{$keyname}[$count])  if($i==1);
					$count++;
				}
			}elsif($mask <= @{$report{$keyname}}){
				push(@newgpu,$report{$keyname}[$mask-1]);
			}
			@{$report{$keyname}} = @newgpu;
		}
	}
}

#==============================================================================

sub runbench(){
	my ($benchname, $report, $keyname, $absorb, $mcxopt)=@_;

	my $mcxoutput=`$MCX --bench $benchname $mcxopt`;
	$mcxoutput=~s/\e\[\d+(?>(;\d+)*)m//g;

	printf("Running benchmark '$MCX --bench $benchname $mcxopt' ... ");

	if($mcxoutput =~ /total simulated energy:/){
		my @lines=split(/\n/,$mcxoutput);
		my %bench=();
		my $tstart=0;
		foreach my $line (@lines){
			if($line =~ /np=(d+)\s+nthread=/){
				$bench{'stat'}{'nphoton'}=$1;
			}elsif($line =~ /simulation speed:\s*([-0-9.]+)/){
				$bench{'speed'}=$1+0;
                                printf(RED);
                                printf("Speed: %.0f photon/ms", $bench{'speed'});
                                printf(RESET);
			}elsif($line =~ /total simulated energy:\s*([-0-9.]+)\s*absorbed: ([-0-9.]+)%/){
				$bench{'stat'}{'energytot'}=$1+0;
				$bench{'stat'}{'absorbfrac'}=$2+0;
				$bench{'stat'}{'energyabs'}=$1*0.01*$2;
			}elsif($line =~ /normalization factor alpha=([-0-9.]+)/){
				$bench{'stat'}{'normalizer'}=$1+0;
			}elsif($line =~ /init complete\s*:\s*([0-9.]+)\s+ms/){
				$tstart=$1;
			}elsif($line =~ /kernel complete:\s+([0-9.]+)\s+ms/){
				$bench{'stat'}{'runtime'}=$1-$tstart;
			}elsif($line =~ /\$Rev::.*(\$Date::[^\$]+\$)/){
				$bench{'mcxversion'}=$1;
			}
		}
		if($mcxopt=~/-W\s+['"]*\s*(([0-9.-]+\s*,\s*)*[0-9.-]+)\s*['"]*/ || $mcxopt=~/--workload\s+['"]*\s*(([0-9.-]+\s*,\s*)*[0-9.-]+)\s*['"]*/){
			$bench{'workload'}="[$1]";
        	}
		if(%bench){
			$report{$keyname}=\%bench;
			my $absfrac=$bench{'stat'}{'absorbfrac'};
			if(abs($absfrac - $absorb)/$absorb > 0.1){
				die('$benchname failed (expected $absorb, got $absfrac), can not continue');
			}
		}else{
			die('$benchname failed, output incomplete');
		}
	}else{
                die('$benchname failed, output incomplete');
	}
        print("\n");
}

#==============================================================================

sub comparegpu(){
	my ($url)=@_;
	my $database;
        $database = `curl --silent $url`;

        if(!defined($database)){
                eval("use LWP::Simple; 1") or warn "LWP::Simple is not available: $@";
                $database = get($url);
		die("fail to download database from $url") if(!defined($database));
	}
	my @res=split(/\n/,$database);
	my $pos=0;
	for my $line (@res){
		if($line =~/<span class='score'>(\d+)</ && $report{'speedsum'}>= $1){
			last;
		}
		$pos++;
	}
	my $wid=`tput cols`;
	if($wid eq ''){
	    $wid=80;
	}

	my $maxscore=0;
	foreach my $id (($pos-5)..($pos+4)){
		next if($id<0 || $id>$#res);
		my $linewid=$wid-37;
		if($id==$pos){
			printf(RED);
			printf("%-25s:%8.0f [%s]\n", '>> Your GPU <<',$report{'speedsum'},'#' x int($linewid*($report{'speedsum'}/$maxscore)));
			printf(RESET);
		}
		if($res[$id]=~/<td>(.*)<\/td><td>\/\s*(.*)\s*\/<br\/><span class='hostid'>(.*)<\/span><\/td><td><div class='barchart'><span class='score'>(\d+)<\/span>/){
			$maxscore=($4>$maxscore)?$4:$maxscore;
			printf("%-25s:%8.0f [%s]\n",substr($2,0,25),$4,'#' x int($linewid*($4/$maxscore)));
		}
	}
}

#==============================================================================

sub submitresult(){
	use JSON::PP;

	my ($report, $url, $ispretty)=@_;
	my %form=();
	my $ans='';
	my %userinfo=(
	  "name"=>"Please provide your full name (will not publish)",
	  "email"=>"Please provide your email (will not publish)[required]",
	  "institution"=>"What is your institution/company (will not publish)?",
	  "nickname"=>"Please provide your web name/ID (visible on webpage)[required]",
	  "machine"=>"Short description of your computer (OS/version/driver, visible)[required]",
	  "comment"=>"Optional comment"
	);
	my @formitem=('name','email','institution','nickname','machine','comment');
	foreach my $key (@formitem){
		print $userinfo{$key}."\n\t";
		$ans=<STDIN>;
		chomp $ans;
		if(($key eq 'nickname' || $key eq 'email' ||  $key eq 'machine' ) && $ans eq ''){
			while($ans eq ''){
				print "required field, can not be empty, please input again\n\t";
				$ans=<STDIN>;
				chomp $ans;
			}
		}
		$report{'userinfo'}{$key}=$ans;
	}
	$form{'name'}=$report{'userinfo'}{'nickname'};
	$form{'time'}= time();
	$form{'ver'}=substr($report{'mcxversion'},7,100);
	$form{'ver'}=~s/\$//g;
	$form{'b1'}=$report{'benchmark1'}{'speed'};
	$form{'b2'}=$report{'benchmark2'}{'speed'};
	$form{'b3'}=$report{'benchmark3'}{'speed'};
	$form{'score'}=$report{'speedsum'};
	$form{'computer'}=$report{'userinfo'}{'machine'}.":/";
	foreach my $gpu (@{$report{'gpu'}}){
		my $gpuname=$gpu->{'name'};
		$gpuname=~s/\s*GeForce\s*//g;
		$gpuname=~s/\s*NVIDIA\s*//g;
		$form{'computer'}.=$gpuname."/";
	}

	my $json = JSON::PP->new;
        $json = $json->pretty($ispretty);

	$form{'report'}= $json->encode($report);

	print "Do you want to see the full data to be submitted ([yes]/no)?";
	$ans=<STDIN>;
	chomp $ans;
	if($ans =~/^yes/i || $ans eq ''){
		print "---------------- begin data -----------------\n";
		print $json->encode(\%form);
		print "\n----------------- end data ------------------\n";
	}
        $ans='';

	print "\nFinal confirmation - please type enter or yes to submit, type no to cancel ([yes]/no)?";
	$ans=<STDIN>;
	chomp $ans;
	if($ans =~/^yes/i || $ans eq ''){
		my $submitdata=$json->encode(\%form);
		my $content=`curl --header 'Content-Type: application/json' --request POST --data '$submitdata' $url`;
		if($content eq ''){
	                eval("use LWP::UserAgent (); 1")  or warn "LWP::UserAgent is not available: $@";;
			my $ua      = LWP::UserAgent->new(); 
			my $response = $ua->post( $url, \%form);
			$content = $response->decoded_content;
			die("fail to submit benchmark data") if($content eq '');
		}
		if($content=~/success/i){
			print "Submission is successful, please browse http://mcx.space/gpubench/ to see the result\n";
		}
	}
}
